internal/graph/singleflight_resolver_test.go

```diff
@@ -0,0 +1,192 @@
+		tuples = append(tuples, tuple.NewTupleKey(fmt.Sprintf("folder:%d", i+1), "parent2", fmt.Sprintf("folder:%d", i)))
+		tuples = append(tuples, tuple.NewTupleKey(fmt.Sprintf("folder:%d", i+1), "parent3", fmt.Sprintf("folder:%d", i)))
+	}
+
+	err := ds.Write(context.Background(), storeID, nil, tuples)
+	require.NoError(t, err)
+
+	typedefs := parser.MustTransformDSLToProto(`model
+	schema 1.1
+
+	type user
+
+	type folder
+	relations
+	  define parent1: [folder]
+	  define parent2: [folder]
+	  define parent3: [folder]
+	  define viewer: [user] or viewer from parent1 or viewer from parent2 or viewer from parent3`).TypeDefinitions
+
+	ctx := typesystem.ContextWithTypesystem(context.Background(), typesystem.New(
+		&openfgav1.AuthorizationModel{
+			Id:              ulid.Make().String(),
+			TypeDefinitions: typedefs,
+			SchemaVersion:   typesystem.SchemaVersion1_1,
+		},
+	))
+
+	checkReq := ResolveCheckRequest{
+		StoreID:            storeID,
+		TupleKey:           tuple.NewTupleKey("folder:10", "viewer", "user:jon"),
+		ContextualTuples:   nil,
+		ResolutionMetadata: &ResolutionMetadata{Depth: defaultResolveNodeLimit},
+	}
+
+	checkerWithoutSingleflight := NewLocalChecker(
+		storagewrappers.NewCombinedTupleReader(ds, []*openfgav1.TupleKey{}),
+	)
+	defer checkerWithoutSingleflight.Close()
+
+	req := checkReq
+	respWithoutSingleflight, err := checkerWithoutSingleflight.ResolveCheck(ctx, &req)
+
+	require.NoError(t, err)
+	require.True(t, respWithoutSingleflight.GetAllowed())
+
+	checkerWithSingleflight := NewLocalChecker(
+		storagewrappers.NewCombinedTupleReader(ds, []*openfgav1.TupleKey{}),
+		WithSingleflightResolver(),
+	)
+	defer checkerWithSingleflight.Close()
+
+	req = checkReq
+	resWithSingleflight, err := checkerWithSingleflight.ResolveCheck(ctx, &req)
+
+	require.NoError(t, err)
+	require.True(t, resWithSingleflight.GetAllowed())
+	require.LessOrEqual(t, resWithSingleflight.GetResolutionMetadata().DatastoreQueryCount, respWithoutSingleflight.GetResolutionMetadata().DatastoreQueryCount)
+}
+
+func TestSingleflightResolverWithCycle(t *testing.T) {
+	ds := memory.New()
+	defer ds.Close()
+
+	storeID := ulid.Make().String()
+
+	var tuples = []*openfgav1.TupleKey{
+		{User: "document:1#viewer3", Relation: "viewer3", Object: "document:1"},
+	}
+
+	err := ds.Write(context.Background(), storeID, nil, tuples)
+	require.NoError(t, err)
+
+	typedefs := parser.MustTransformDSLToProto(`model
+	schema 1.1
+  
+  type user
+  
+  type document
+	relations
+		define viewer1: [user, document#viewer1]
+		define viewer2: viewer1 or viewer2
+	  	define viewer3: viewer1 or viewer2`).TypeDefinitions
+
+	ctx := typesystem.ContextWithTypesystem(context.Background(), typesystem.New(
+		&openfgav1.AuthorizationModel{
+			Id:              ulid.Make().String(),
+			TypeDefinitions: typedefs,
+			SchemaVersion:   typesystem.SchemaVersion1_1,
+		},
+	))
+
+	checkerWithoutSingleflight := NewLocalChecker(
```

Comment:
This name is misleading as we are passing in `WithSingleflightResolver()`, but the name implies we're not.